internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -179,6 +189,112 @@ func getOwnerReferenceFrom(obj, owner client.Object) *metav1.OwnerReference {
 	return nil
 }
 
+func (r *Reconciler) callAfterHooks(ctx context.Context, s *scope.Scope) error {
+	if err := r.callAfterControlPlaneInitialized(ctx, s); err != nil {
+		return err
+	}
+
+	if err := r.callAfterClusterUpgrade(ctx, s); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) callAfterControlPlaneInitialized(ctx context.Context, s *scope.Scope) error {
+	/*
+		TODO: Working comment - DELETE AFTER:
+		- If the cluster topology is being created then mark the AfterControlPlaneInitialized hook so that we can call it later.
+	*/
+	if s.Current.Cluster.Spec.InfrastructureRef == nil && s.Current.Cluster.Spec.ControlPlaneRef == nil {
+		if err := hooks.MarkAsPending(ctx, r.Client, s.Current.Cluster, runtimehooksv1.AfterControlPlaneInitialized); err != nil {
+			return errors.Wrapf(err, "failed to mark %s hook", runtimecatalog.HookName(runtimehooksv1.AfterControlPlaneInitialized))
+		}
+	}
+
+	if hooks.IsPending(runtimehooksv1.AfterControlPlaneInitialized, s.Current.Cluster) {
+		if isControlPlaneInitialized(s.Current.Cluster) {
+			hookRequest := &runtimehooksv1.AfterControlPlaneInitializedRequest{
+				Cluster: *s.Current.Cluster,
+			}
+			hookResponse := &runtimehooksv1.AfterControlPlaneInitializedResponse{}
+			if err := r.RuntimeClient.CallAllExtensions(ctx, runtimehooksv1.AfterControlPlaneInitialized, s.Current.Cluster, hookRequest, hookResponse); err != nil {
+				return errors.Wrapf(err, "failed to call %s hook", runtimecatalog.HookName(runtimehooksv1.AfterControlPlaneInitialized))
+			}
+			s.HookResponseTracker.Add(runtimehooksv1.AfterControlPlaneInitialized, hookResponse)
+			if err := hooks.MarkAsDone(ctx, r.Client, s.Current.Cluster, runtimehooksv1.AfterControlPlaneInitialized); err != nil {
+				return errors.Wrapf(err, "failed to unmark %s hook", runtimecatalog.HookName(runtimehooksv1.AfterControlPlaneInitialized))
+			}
+		}
+	}
+
+	return nil
+}
+
+func isControlPlaneInitialized(cluster *clusterv1.Cluster) bool {
+	for _, condition := range cluster.GetConditions() {
+		// TODO: Should we check for the ControlPlaneInitialized condition or the ControlPlaneReadyCondition?
+		// From the description of the hook it looks like it should be the ControlPlaneReadyCondition - but need to double check.
```

Comment:
Could there be the use-case that an `AfterControlPlaneIntialized` hook installs CNI? If so, the `ControlPlaneReadyCondition` could only be true after this is the case.

Reply:
As written in a previous comment. I lean towards `ControlPlaneInitialized` given its definition:
> // ControlPlaneInitializedCondition reports if the cluster's control plane has been initialized such that the
> // cluster's apiserver is reachable and at least one control plane Machine has a node reference.